src/utilities/crypto.cc

```diff
@@ -275,72 +269,56 @@ bool Crypto::generateRSAKeyPair(KeyType key_type, std::string *public_key, std::
   int ret = 0;
 
 #if AKTUALIZR_OPENSSL_PRE_11
-  RSA *r = RSA_generate_key(bits,   /* number of bits for the key - 2048 is a sensible value */
-                            RSA_F4, /* exponent - RSA_F4 is defined as 0x10001L */
-                            NULL,   /* callback - can be NULL if we aren't displaying progress */
-                            NULL    /* callback argument - not needed in this case */
-                            );
+  StructGuard<RSA> rsa(RSA_generate_key(bits,   /* number of bits for the key - 2048 is a sensible value */
+                                        RSA_F4, /* exponent - RSA_F4 is defined as 0x10001L */
+                                        NULL,   /* callback - can be NULL if we aren't displaying progress */
+                                        NULL    /* callback argument - not needed in this case */
+                                        ),
+                       RSA_free);
 #else
-  BIGNUM *bne = BN_new();
-  ret = BN_set_word(bne, RSA_F4);
+  StructGuard<BIGNUM> bne(BN_new(), BN_free);
+  ret = BN_set_word(bne.get(), RSA_F4);
   if (ret != 1) {
-    BN_free(bne);
     return false;
   }
-  RSA *r = RSA_new();
-  ret = RSA_generate_key_ex(r, bits, /* number of bits for the key - 2048 is a sensible value */
-                            bne,     /* exponent - RSA_F4 is defined as 0x10001L */
-                            NULL     /* callback argument - not needed in this case */
+  StructGuard<RSA> rsa(RSA_new(), RSA_free);
+  ret = RSA_generate_key_ex(rsa.get(), bits, /* number of bits for the key - 2048 is a sensible value */
+                            bne.get(),       /* exponent - RSA_F4 is defined as 0x10001L */
+                            NULL             /* callback argument - not needed in this case */
                             );
   if (ret != 1) {
-    RSA_free(r);
-    BN_free(bne);
     return false;
   }
 #endif
 
-  EVP_PKEY *pkey = EVP_PKEY_new();
-  EVP_PKEY_assign_RSA(pkey, r);
+  StructGuard<EVP_PKEY> pkey(EVP_PKEY_new(), EVP_PKEY_free);
+  // release the rsa pointer here, pkey is the new owner
+  EVP_PKEY_assign_RSA(pkey.get(), rsa.release());
   char *pubkey_buf;
-  BIO *pubkey_sink = BIO_new(BIO_s_mem());
+  StructGuard<BIO> pubkey_sink(BIO_new(BIO_s_mem()), BIO_vfree);
   if (!pubkey_sink) {
     return false;
   }
-  ret = PEM_write_bio_PUBKEY(pubkey_sink, pkey);
+  ret = PEM_write_bio_PUBKEY(pubkey_sink.get(), pkey.get());
   if (ret != 1) {
-    RSA_free(r);
-#if AKTUALIZR_OPENSSL_AFTER_11
-    BN_free(bne);
-#endif
```

Comment:
Is this logic definitely not needed any more?  I can't see where it moved to. I think @OYTIS understands the reasons for it.

Reply:
Ah, OK. I'd missed that